Velocity pruning

Jmeyer1292 opened the pull request:
Related to issue #63 and Issue #48, here is my first crack at adding time windowing data to our trajectory points. First off, this pull-request is not quite ready because there are many documentation changes that need to be made but I would like to get eyes on this asap.

This PR replaces the Point Transition class with a timing constraint struct that has an upper and lower limit for time. Only the upper limit on time is currently used or checked (I'm not sure if the lower is needed). This timing must be specified when building new types of Trajectory Points but it can be default constructed which puts the upper and lower limits to zero.

Planning graph has been changed to incorporate pull request #64 as well as to remove the absolute joint delta check in place of an updated function in RobotModel that takes two joint positions and a time delta and sees if motion between those two points is possible in that window of time.

Again this is under development, but I want to see if people think I'm on the track.

Commit:
First take on creating a timing constraint element for plans

Commit:
Made FindStartVertices and FindEndVertices go off of the users input. Also actually removed edges when they had high weight.

ClayFlannigan commented:
Something for us to consider: a fully specified joint state would specify the position and velocity for each point.  Acceptable trajectories would require less than than maximum joint velocities and accelerations between points.  The current planner only considers positions and ignores dynamics (velocities and accelerations).  This pull adds time constraints thereby permitting calculation of average velocities along a trajectory.  In the limit as the step size approaches zero, the average and the instantaneous velocities are the same.  The question is: should we instead specify velocities at points, thereby permitting us to calculate accelerations to achieve the trajectory?  Or should we look at the second (numerical) derivative to calculate accelerations?  

UPDATE: I think the above is flawed, because we can't know the joint velocities ahead of time.  All we know is time, so calculating the average velocity is the best we can do.  I guess there is still the question of whether we should also calculate the average acceleration?

gavanderhoorn commented:
Perhaps also something to keep in mind (or to purposefully ignore): a lot (some?) industrial robot controllers work with trajectory segments, where you can specify a target velocity to be reached over that segment, not at the end point. With blending/falling through/contouring/x, you can get the controller to smoothly stitch segments together, but at the cost of accuracy (in both time and space, I believe). Making the segments small enough can work around this in some way, but (in most cases) at the cost of maximum attainable performance.

I've always considered this more position control with velocity constraints over traj segments than actually being able to define the state of the manipulator (pose & velocity) at specific points. ROS generated trajectories seem to assume we can actually do the latter. I think only the `motoman` driver is currently capable of that within ros-i.

Is this something we want to keep in mind? Or is `descartes` being designed _for the future_, where all industrial controllers provide full access to their motion control systems?

ClayFlannigan commented:
I agree with @gavanderhoorn in that it is probably best to consider it as position control with velocity constraints and rather than a true velocity controller, so the current approach seems like the pragmatic and right one.  

Related, a filter could be applied post-planning that generates a smooth velocity trajectory.  Here is the classic cubic spline equation which requires as inputs pos_0, pos_1, t_0, t_1, and vel_0 to calculate the smooth velocity at the trajectory endpoint, vel_1. A similar approach could be used if we wanted to specify joint velocities.

from: Spong et. al., "Robot Modeling and Control", 2005
![cubic_interpolated_trajectory](https://cloud.githubusercontent.com/assets/6444485/6545889/b931483e-c56a-11e4-863e-570215d7a33d.jpg)

shaun-edwards commented on descartes_core/include/descartes_core/trajectory_timing_constraint.h:
```diff
@@ -0,0 +1,52 @@
+ * trajectory_pt_transition.h
+ *
+ *  Created on: Jun 5, 2014
+ *      Author: Dan Solomon
```
Update author.

shaun-edwards commented on descartes_moveit/include/descartes_moveit/moveit_state_adapter.h:
```diff
@@ -48,7 +47,8 @@ class MoveitStateAdapter : public descartes_core::RobotModel
   MoveitStateAdapter(const moveit::core::RobotState & robot_state, const std::string & group_name,
-                    const std::string & tool_frame, const std::string & world_frame,
+                     const std::string & tool_frame, const std::string & world_frame,
+                     const std::vector<double>& velocity_limits,
```
Aren't the velocity limits stored in the `moveit::core::RobotState`?

shaun-edwards commented on descartes_core/include/descartes_core/robot_model.h:
```diff
@@ -102,6 +102,17 @@ class RobotModel
+   * @param  dt              [description]
+   * @return                 [description]
+   */
+  virtual bool isValidMove(const std::vector<double>& from_joint_pose, const std::vector<double>& to_joint_pose,
```
Should this method name be more specific?  It really only checks velocity limits?  The name implies more.

Jmeyer1292 replied:
I was originally thinking that the method might be a general test for the validity of a motion and the specifics would be up to the implementation. That way we could incorporate acceleration or collisions, for example.

Admittedly that's probably not going to happen, so would something like checkVelocityConstraints() more apt?

Jmeyer1292 replied:
ping

shaun-edwards commented on descartes_moveit/src/moveit_state_adapter.cpp:
```diff
@@ -33,13 +33,15 @@ namespace descartes_moveit
 MoveitStateAdapter::MoveitStateAdapter()
 {
-
+  velocity_limits_ = std::vector<double>(7, 0.1);
```
This should be removed if the MoveIt robot state already contains this info.

shaun-edwards commented on descartes_planner/src/planning_graph.cpp:
```diff
@@ -41,7 +41,7 @@ using namespace descartes_trajectory;
 {
 
-const double MAX_JOINT_DIFF = M_PI;
+const double MAX_JOINT_DIFF = M_PI / 20;
```
This can go away now...right?

shaun-edwards commented on descartes_planner/src/planning_graph.cpp:
```diff
@@ -595,39 +595,81 @@ bool PlanningGraph::removeTrajectory(TrajectoryPtPtr point)
+  // Test for error case
+  if (cart_id.is_nil())
+  {
+    ROS_ERROR("Could not locate TrajectoryPt with nil previous point. Graph may be cyclic.");
```
Should this ever happen?  Maybe a comment to describe when it does (if it could)?

Jmeyer1292 replied:
The addTrajectory function of planning_graph takes a pointer and a previous and next id. One could add arbitrary edges with this.

shaun-edwards commented on descartes_planner/src/planning_graph.cpp:
```diff
@@ -595,39 +595,81 @@ bool PlanningGraph::removeTrajectory(TrajectoryPtPtr point)
+  // Test for error case
+  if (cart_id.is_nil())
+  {
+    ROS_ERROR("Could not locate TrajectoryPt with nil next point. Graph may be cyclic.");
```
See comment above.

shaun-edwards commented on descartes_planner/src/planning_graph.cpp:
```diff
@@ -1004,19 +1058,17 @@ double PlanningGraph::linearWeight(JointTrajectoryPt start, JointTrajectoryPt en
     {
+      if (end.getTiming().isSpecified() && !robot_model_->isValidMove(start_vector, end_vector, end.getTiming().upper_))
+      {
+        return MAX_EXCEEDED_PENALTY;
```
Am I reading this correctly...if the timing is specified and the move is not valid, we set the max penalty?  Shouldn't we prune the edge?  Or is this a "magic" value that indicates we should prune the edge?

shaun-edwards commented on descartes_core/include/descartes_core/trajectory_timing_constraint.h:
```diff
@@ -0,0 +1,52 @@
+  /**
+   * Timing constraints 
+   */
+  struct TimingConstraint
```
Added constructor for a timing constraint with a single value (i.e. just a nominal timing).

shaun-edwards commented:
@gavanderhoorn, @ClayFlannigan, it sounds like you are OK with the current approach.  I would like provide a little bit of the total vision (this will eventually make it's way to the REP).  When considering trajectories, Descartes will contain two types of trajectories:
- Dense trajectory (currently implemented) - A dense trajectory is defined point by point.  A trajectory point (joint/cart/whatever) will contain a pose, with associated velocity, acceleration, and timing.  I realize this could be over/under constrained depending how many derivatives you think are required.  A key requirement is that motion in between trajectory points is well defined (i.e. smooth and collision free).
- Sparse trajectory (in progress, see start in PR #51) - This could be any number of representations.  It could look like traditional robot code (as @gavanderhoorn pointed out), gcode, or anything.  In general, these trajectories will contain points and segment constrains (linear, circular, joint, velocity, timing, etc...).  These representations are efficient, but there is no guarantee that motion segments are well behaved and/or is collision free.  Sparse trajectories must be converted to dense trajectories for planning.  This conversion depends on the representation used, but should result in populated velocity, acceleration, and timing values.

Initially @Jmeyer1292 and I chose timing because it generalizes to any kind of point.  Therefore, it can exist in the base `TrajectoryPt` type.  I believe this was Clay's conclusion as well.  The one possible downside to this change implies a direction to the motion (i.e. a `delta t` from the previous point).  A `delta t` approach was chosen to allow an upper and lower bound to be specified (kind of like the pose constraints).  Not sure how to avoid the implied direction of motion.

To @ClayFlannigan 's second point, the joint trajectory that results from planning will have to be smoothed.  If the graph used for planning does not contain edges or nodes that are physically impossible, then a smooth trajectory through the planned trajectory is likely possible.  Unfortunately, there are some trajectory conditions, like path reversals, which are not considered at pruning time.  This is a known limitation at this time.

One final thought, @ClayFlannigan and I discussed using instantaneous velocities at a point to determine if the point (and therefore the node) is valid.  Using this PR's method of pruning edges is good, but removing a node entirely is much better.  I believe I understand how to generalize this to both cart/joint points.  This will be a future PR.

Commit:
Updated the TimingConstraint impl with documentation and a new nominal time constructor

Commit:
Reworked linearWeight to return a pair of success,cost and removed magic constants

Commit:
Added getJointVelocityLimits function to moveit

Commit:
The non-default constructor of MoveItStateAdapter and the initialize method now attempt to find the velocity limits from moveit

Commit:
Added comments explaining cyclic graph issue

madratman commented:
For smoothing, can [this](http://wiki.ros.org/trajectory_filters) be useful?
[More details in fourth comment in issue #52](https://github.com/ros-industrial-consortium/descartes/issues/52)

Commit:
Added timing constraints to constructors of AxialSymmetricPt

shaun-edwards commented:
@Jmeyer1292, this PR can't be merged automatically...can you rebase to fix this?

Commit:
Fixed removed some unncessary headers and objects related to UUID usage

Jmeyer1292 commented:
@shaun-edwards I rebased the code, but there is still the outstanding issue of making the timing work with the sparse trajectory planner. I'll look into that shortly.

Commit:
Updated sparse planner to take timing info into account

Commit:
Fixed descartes_trajectory unit tests and added some assurances for time values

Jmeyer1292 commented:
@shaun-edwards or @jrgnicho If y'all would take a look at this again when you have a chance, I'd appreciate it.

shaun-edwards commented:
@jrgnicho, please review

jrgnicho commented on descartes_core/include/descartes_core/trajectory_timing_constraint.h:
```diff
@@ -0,0 +1,80 @@
+     */
+    bool isSpecified() const { return upper_ > 0.0; }
+
+    double lower_, upper_;
```
Struct naming convention [see here](http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Variable_Names)

jrgnicho commented on descartes_moveit/src/moveit_state_adapter.cpp:
```diff
@@ -315,5 +357,29 @@ int MoveitStateAdapter::getDOF() const
+  }
+
+  return true;
+}
```
Maybe we should add a check for same size arrays

Jmeyer1292 replied:
What's our policy on using asserts?

jrgnicho replied:
We don't really have a policy for this but I suppose that it is justifiable if the application runs into an irrecoverable condition.  My preference is printing a ROS_ERROR and returning false.

jrgnicho commented on descartes_planner/src/dense_planner.cpp:
```diff
@@ -97,6 +96,7 @@ bool DensePlanner::updatePath()
       || (std::find_if(cart_map.begin(),cart_map.end(),predicate) == cart_map.end())
       || first_id == descartes_core::TrajectoryID::make_nil())
   {
+    ROS_ERROR_STREAM("Failure case 1");
```
What is a "Failure case 1" error?

jrgnicho commented on descartes_planner/src/dense_planner.cpp:
```diff
@@ -137,11 +138,13 @@ bool DensePlanner::updatePath()
   {
+    ROS_ERROR_STREAM("Failure case 3");
     error_code_ = descartes_core::PlannerError::IK_NOT_AVAILABLE;
   }
```
Can the error message be made more specific to the type of failure?

jrgnicho commented on descartes_planner/src/planning_graph.cpp:
```diff
@@ -672,7 +675,7 @@ bool PlanningGraph::getShortestPath(double &cost, std::list<JointTrajectoryPt> &
     for (std::list<JointGraph::vertex_descriptor>::iterator end = end_points.begin(); end != end_points.end(); end++)
     {
-      //ROS_INFO("Checking path: S[%d] -> E[%d]", *start, *end);
+      // ROS_INFO("Checking path: S[%d] -> E[%d]", *start, *end);
```
This should be turned into a ROS_DEBUG .. or be removed

jrgnicho commented on descartes_planner/src/sparse_planner.cpp:
```diff
@@ -730,6 +758,22 @@ int SparsePlanner::interpolateSparseTrajectory(const SolutionArray& sparse_solut
+          double dt = timing_cache_[pos].upper_;
+          // check validity of joint motion
+
+          if (!robot_model->isValidMove(sol, aprox_interp, dt))
```
I think this **isValidMove** check makes the **checkJointChanges** call above unnecessary.

Jmeyer1292 replied:
Well, it seems to me that `checkJointChanges` tests the interpolated joint position against the valid IK for that point. `isValidMove` checks to see if that point is achievable from the previous one given the joint velocity constraints of the system.

Commit:
Renamed variables in timing constraint to remove trailing underscores. Various other small fixups.

jrgnicho commented on descartes_core/include/descartes_core/trajectory_timing_constraint.h:
```diff
@@ -0,0 +1,80 @@
+     * @brief Checks if the given timing constraint has been specified
+     * @return true if the point is specified
+     */
+    bool isSpecified() const { return upper > 0.0; }
```
Should we check for negative values on the **lower** member as well?

Jmeyer1292 replied:
I'm more inclined to put such a check in as a ROS_ASSERT into the constructor. I did not mean this function as a validity check but a test for whether or not the user actually specified a dt.

jrgnicho commented on descartes_moveit/src/moveit_state_adapter.cpp:
```diff
@@ -26,6 +26,37 @@
+    const auto& bounds = model->getVariableBounds();
+    // Check to see if there is a single bounds constraint (more might indicate
+    // not revolute joint)
+    if (bounds.size() != 1)
```
I think it be more reliable to check the [type](http://docs.ros.org/indigo/api/moveit_core/html/classmoveit_1_1core_1_1JointModel.html#a17a4c66425888df27c53c84bb1fa0585) to determine if they are either revolute or prismatic joints.

Jmeyer1292 replied:
+1 I did that on the IK seed generator, but wasn't aware of it when I started with this

jrgnicho commented on descartes_moveit/src/moveit_state_adapter.cpp:
```diff
@@ -320,5 +363,36 @@ int MoveitStateAdapter::getDOF() const
+  // Check for equal sized arrays
+  if (from_joint_pose.size() != to_joint_pose.size())
+  {
+    ROS_DEBUG_STREAM("To and From joint poses are of different sizes.");
```
This should probably be a ROS_ERROR

Jmeyer1292 replied:
+1

jrgnicho commented on descartes_planner/src/sparse_planner.cpp:
```diff
@@ -23,11 +23,28 @@
+    }
+}
+
+
```
Would it not be a bit more efficient to allow for an output argument for the vector instead of returning a copy?

Jmeyer1292 replied:
I hesitate to make general comments on compiler performance without benchmarking, but I doubt that output arguments are any faster because of return-value-optimizations and the fact that in c++-11 values returned from functions are implicitly moved.

jrgnicho commented on descartes_planner/src/sparse_planner.cpp:
```diff
@@ -131,6 +148,15 @@ bool SparsePlanner::planPath(const std::vector<TrajectoryPtPtr>& traj)
+
+  // Save the timing information
+  timing_cache_ = extractTiming(traj);
+  // Unconstrian the times for the sparse sampling
```
typo Unconstrian

Jmeyer1292 replied:
+1

jrgnicho commented:
Code reviews aren't fun.  Just a few minor comments but in general this one is ready to merge

Commit:
Changed the getJointVelocityLimits function to check the type of the joint instead of the bounds size()

Commit:
fixed typo

Commit:
Added assert to timing constraint that enforces that all lower and upper thresholds should be positive for the moment

Commit:
Replaced the assert with a series of timing checks which print warning messages and set the time to be unspecified

Jmeyer1292 commented:
@jrgnicho I removed the assertion and replaced it with a warning. In the event of invalid (read negative) timing values, the times are instead set to zero indicating unspecified.

jrgnicho commented:
+1
